core/src/main/scala/org/apache/spark/ui/jobs/StageTable.scala

```diff
@@ -145,9 +145,20 @@ private[ui] class StageTableBase(
       case None => "Unknown"
     }
     val finishTime = s.completionTime.getOrElse(System.currentTimeMillis)
-    val duration = s.submissionTime.map { t =>
-      if (finishTime > t) finishTime - t else System.currentTimeMillis - t
-    }
+    val taskLaunchTimes = stageData.taskData.values
+        .filter(_.taskInfo.launchTime != 0)
+        .map(_.taskInfo.launchTime)
```

Comment:
to improve readability, can you do the following instead
```
val firstTaskLaunchTime = stageData.taskData.values
  .filter(...)
  .map(...)
  .min
```
then add a comment saying how the stage's submission time is misleading because it counts all the time the stage spends waiting to be launched (SPARK-10930).